File: easy/knight_moves.pl
use strict;
open( INFILE, "<", $ARGV[0] )
  or die("Cannot open file $ARGV[0] for reading: $!");
while ( my $line = <INFILE> ) {
    my ( $l, $r ) = ( ord $line, substr( $line, 1, 1 ) );
    my @s;
    push @s, chr( $l - 2 ) . ( $r - 1 ) if $l > ord "b" && $r > 1;
    push @s, chr( $l - 2 ) . ( $r + 1 ) if $l > ord "b" && $r < 8;
    push @s, chr( $l - 1 ) . ( $r - 2 ) if $l > ord "a" && $r > 2;
    push @s, chr( $l - 1 ) . ( $r + 2 ) if $l > ord "a" && $r < 7;
    push @s, chr( $l + 1 ) . ( $r - 2 ) if $l < ord "h" && $r > 2;
    push @s, chr( $l + 1 ) . ( $r + 2 ) if $l < ord "h" && $r < 7;
    push @s, chr( $l + 2 ) . ( $r - 1 ) if $l < ord "g" && $r > 1;
    push @s, chr( $l + 2 ) . ( $r + 1 ) if $l < ord "g" && $r < 8;
    printf( "%s\n", join( ' ', @s ) );
}
close(INFILE);

File: easy/knight_moves.rb
def pos(l, r)
  (l + 'a'.ord).chr + (r + '1'.ord).chr
end

File.open(ARGV[0]).each_line do |line|
  l = line[0].ord - 'a'.ord
  r = line[1].ord - '1'.ord
  m = []
  m << pos(l - 2, r - 1) if l > 1 && r > 0
  m << pos(l - 2, r + 1) if l > 1 && r < 7
  m << pos(l - 1, r - 2) if l > 0 && r > 1
  m << pos(l - 1, r + 2) if l > 0 && r < 6
  m << pos(l + 1, r - 2) if l < 7 && r > 1
  m << pos(l + 1, r + 2) if l < 7 && r < 6
  m << pos(l + 2, r - 1) if l < 6 && r > 0
  m << pos(l + 2, r + 1) if l < 6 && r < 7
  puts m.join(' ')
end
